song_scrounger/document_parser.py

```diff
@@ -1,49 +1,128 @@
-
-def _find_first_two_quotes(text):
-    """Finds indices of first two quotation marks.
-
-    e.g. 'A "quote"' => (2,8)
-    e.g. 'No quote' => None
-    e.g. 'Not "balanced' => None
-
-    Params:
-        text (str): e.g. 'A "quote"'.
-    Returns:
-        ((int,int)): indices of first two quotes in given text. None if absent or unbalanced.
-    """
-    if len(text) <= 1:
-        return None
-
-    opening_quote_index = text.find("\"")
-    if opening_quote_index != -1:
-        closing_quote_index = text.find("\"", opening_quote_index+1)
-        if closing_quote_index != -1:
-            return opening_quote_index, closing_quote_index
-    return None
\ No newline at end of file
+from .spotify_client import SpotifyClient
+from .models.song import Song
+
+
+class DocumentParser():
+    def __init__(self, spotify_client):
+        self.spotify_client = spotify_client
+
+    async def find_songs(self, text):
+        """Parses given text for songs, matching with artists if mentioned.
+
+        Each song is searched on Spotify. The artists in the search results
+        are searched for in the text as well. Any matches are used for
+        song disambiguation.
+
+        Params:
+            text (str): multi-paragraph page containing song names
+                and perhaps some of their artists.
+
+        Returns:
+            (dict): key (str) is song name; val ([str]) is spotify URIs
+                of matching songs, empty if no matching artist mentioned.
+        """
+        results = {}
+        paragraphs = self._get_paragraphs(text)
+        for paragraph in paragraphs:
+            song_names = self.find_quoted_tokens(paragraph)
+            for song_name in song_names:
+                songs = await self.search_spotify(song_name)
+                songs_with_mentioned_artists = self.filter_by_mentioned_artist(songs, text)
+                if len(songs_with_mentioned_artists) > 0:
+                    results[song_name] = [song.spotify_uri for song in songs_with_mentioned_artists]
+                else:
+                    results[song_name] = [song.spotify_uri for song in songs]
+        return results
+
+    def filter_by_mentioned_artist(self, songs, text):
+        songs_whose_artists_are_mentioned = set()
+        for song in songs:
+            for artist in song.artists:
+                if self.is_mentioned(artist, text):
+                    songs_whose_artists_are_mentioned.add(song)
+        return songs_whose_artists_are_mentioned
+
+    async def search_spotify(self, song):
+        """
+        Params:
+            song (str): e.g. "Sorry".
+
+        Returns:
+            songs ([Song]).
+        """
+        songs = []
+        tracks = await self.spotify_client.find_track(song)
+        for track in tracks:
+            for artist in track.artists:
+                songs.append(Song(
+                    track.name,
+                    track.uri,
+                    [artist.name for artist in track.artists]
```

Comment:
Something is not right here...  It looks like you're just adding |track.artists| number of Song objects to the songs list, but you're making it O(|track.artists|^2) by doing that list comprehension in the constructor.

At the very least you could store the list in a variable outside the artist in track.artists loop so you could re-use it (or you could write a function which returns a list of identical Song objects from a track based on the number of artists on the track object), but I'm curious as to why you're adding duplicates here anyway.

Reply:
Yeah the logic is busted here, I've fixed it locally and pushed up the changes.